include/rx/ranges.hpp

```diff
@@ -654,107 +831,151 @@ struct transform {
                 input_.next();
-            } while (!input_.at_end() && !filter_.pred(input_.get()));
+            }
         }
+
         [[nodiscard]] constexpr bool at_end() const noexcept {
             return input_.at_end();
         }
+
         [[nodiscard]] constexpr size_t size_hint() const noexcept {
             return input_.size_hint();
         }
     };
 
+    // Since F may be either cheap or expensive to copy, provide overloads for both const-refs and
+    // rvalue-refs.
+
     template <class InputRange>
     [[nodiscard]] constexpr auto operator()(InputRange&& input) const& {
-        using Inner = RX_REMOVE_CVREF_T<decltype(as_input_range(std::forward<InputRange>(input)))>;
-        return Range<Inner>{as_input_range(std::forward<InputRange>(input)), *this};
+        auto inner = as_idempotent_input_range(std::forward<InputRange>(input));
+        using Inner = decltype(inner);
+        return Range<Inner>{std::move(inner), pred};
     }
     template <class InputRange>
     [[nodiscard]] constexpr auto operator()(InputRange&& input) && {
-        using Inner = RX_REMOVE_CVREF_T<decltype(as_input_range(std::forward<InputRange>(input)))>;
-        return Range<Inner>{as_input_range(std::forward<InputRange>(input)), std::move(*this)};
+        auto inner = as_idempotent_input_range(std::forward<InputRange>(input));
+        using Inner = decltype(inner);
+        return Range<Inner>{std::move(inner), std::move(pred)};
     }
 };
 template <class F>
-filter(F &&)->filter<RX_REMOVE_CVREF_T<F>>;
+filter(F &&)->filter<remove_cvref_t<F>>;
 
 /*!
     @brief Create a range that produces at most N elements from the beginning of its input.
 */
 struct take {
-    size_t n;
+    const size_t n;
     constexpr explicit take(size_t n) noexcept : n(n) {}
 
-    template <class InputRange>
+    template <class R>
     struct Range {
-        InputRange input_;
-        size_t i = 0;
-        size_t n;
-        using output_type = typename InputRange::output_type;
+        using output_type = typename R::output_type;
         static constexpr bool is_finite = true;
+        static constexpr bool is_idempotent = is_idempotent_v<R>;
+
+        R inner;
+        const size_t n;
+        size_t i = 0;
+
+        constexpr Range(R inner, size_t n) noexcept : inner(std::move(inner)), n(n) {}
 
-        constexpr Range(InputRange input, size_t n) noexcept : input_(std::move(input)), n(n) {}
         [[nodiscard]] constexpr output_type get() const noexcept {
-            return input_.get();
+            return inner.get();
         }
+
         constexpr void next() noexcept {
-            RX_ASSERT(!input_.at_end());
+            RX_ASSERT(!inner.at_end());
             ++i;
-            input_.next();
+            inner.next();
         }
+
         [[nodiscard]] constexpr bool at_end() const noexcept {
-            return i >= n || input_.at_end();
+            return i == n || inner.at_end();
```

Comment:
Is this construct too ugly? It could aid ranges where calling at_end() is at least a little expensive:

```c++
if (i == n) {
    return true;
} else if (inner.at_end()) {
    i = n;
    return true;
} else {
    return false;
}
```

Reply:
Hm yes, but that would require making `i` mutable.

I think many things rely on `at_end()` being cheap already, which is also the rationale for making it `noexcept`.